swap/int256/int256_test.go

```diff
@@ -0,0 +1,191 @@
+package int256
+
+import (
+	"io/ioutil"
+	"math/big"
+	"testing"
+
+	"github.com/ethersphere/swarm/state"
+)
+
+type testCase struct {
+	name         string
+	value        *big.Int
+	expectsError bool
+}
+
+// the following test cases cover a range of values to be used to create int256 variables
+// these variables are expected to be created successfully when using integer values
+// contained in the closed interval between -2^255 and 2^255 - 1
+var int256TestCases = []testCase{
+	{
+		name:         "case 0",
+		value:        big.NewInt(0),
+		expectsError: false,
+	},
+	// negative numbers
+	{
+		name:         "case -1",
+		value:        big.NewInt(-1),
+		expectsError: false,
+	},
+	{
+		name:         "case -1 * 2^8",
+		value:        big.NewInt(-256),
+		expectsError: false,
+	},
+	{
+		name:         "case -1 * 2^64",
+		value:        new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(64), nil)),
+		expectsError: false,
+	},
+	{
+		name:         "case -1 * 2^255",
+		value:        new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil)),
+		expectsError: false,
+	},
+	{
+		name:         "case -1 * 2^255 - 1",
+		value:        new(big.Int).Sub(new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil)), big.NewInt(1)),
+		expectsError: true,
+	},
+	{
+		name:         "case -1 * 2^512",
+		value:        new(big.Int).Mul(big.NewInt(-1), new(big.Int).Exp(big.NewInt(2), big.NewInt(512), nil)),
+		expectsError: true,
+	},
+	// positive numbers
+	{
+		name:         "case 1",
+		value:        big.NewInt(1),
+		expectsError: false,
+	},
+	{
+		name:         "case 2^8",
+		value:        big.NewInt(256),
+		expectsError: false,
+	},
+	{
+		name:         "case 2^128",
+		value:        new(big.Int).Exp(big.NewInt(2), big.NewInt(128), nil),
+		expectsError: false,
+	},
+	{
+		name:         "case 2^255 - 1",
+		value:        new(big.Int).Sub(new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil), big.NewInt(1)),
+		expectsError: false,
+	},
+	{
+		name:         "case 2^255",
+		value:        new(big.Int).Exp(big.NewInt(2), big.NewInt(255), nil),
+		expectsError: true,
+	},
+	{
+		name:         "case 2^512",
+		value:        new(big.Int).Exp(big.NewInt(2), big.NewInt(512), nil),
+		expectsError: true,
+	},
+}
+
+// TestSet tests the creation of valid and invalid Int256 structs by calling the Set function
+func TestInt256Set(t *testing.T) {
+	for _, tc := range int256TestCases {
+		t.Run(tc.name, func(t *testing.T) {
+			result, err := NewInt256(*tc.value)
+			if tc.expectsError && err == nil {
+				t.Fatalf("expected error when creating new Int256, but got none")
+			}
+			if !tc.expectsError {
+				if err != nil {
+					t.Fatalf("got unexpected error when creating new Int256: %v", err)
+				}
+				resultValue := result.Value()
+				if (&resultValue).Cmp(tc.value) != 0 {
+					t.Fatalf("expected value of %v, got %v instead", tc.value, result.value)
+				}
+			}
+		})
+	}
+}
+
```

Comment:
I actually added another test here to see if value manipulation is possible through the `Value()` getter and it is:
```go
func TestIn256ValueInjection(t *testing.T) {
	i := new(big.Int).SetInt64(12)
	v, err := NewInt256(*i)
	if err != nil {
		t.Fatalf("got unexpected error when creating new Int256: %v", err)
	}

	vv := v.Value()
	vvc := v.Value()

	vv.SetInt64(15)

	if vvc.Cmp(&vv) == 0 {
		t.Fatal("values are equal though one should change: %v %v", vv, vvc)
	}
}
```

Reply:
thanks for pointing this out, you're correct.

i'm going to keep this test while i work on the pointer value for `Int256`

Reply:
i changed some numbers and noticed this:
```
v := big.NewInt(-3)
vp := *v
	
fmt.Println(v)  // prints -3
	
vp.SetInt64(5) 
	
fmt.Println(v) // prints -5, not 5
```

am i missing how this makes sense?

Reply:
This is quite interesting, but actually to mutate an instance, as SetInt64 is doing it, it has to be called on a pointer, which is not the case here. The sign is not changed because of that, but the value is, as it is implemented using a slice.

In this example, only the returned `nvp` has the correct value, as it is returned as mutated `vp`.

```
	v := big.NewInt(-3)
	vp := *v

	fmt.Println(v) // prints -3

	nvp := vp.SetInt64(5)

	fmt.Println(v) // prints -5, not 5
	
	fmt.Println(nvp) // prints 5
```

I think that it is important to keep the expected semantics which is in this case pointer semantics when calling methods.

There is a short explanation about shallow copies of Int type https://golang.org/pkg/math/big/#Int.

Reply:
very cool @janos, thanks!

Reply:
this is due to the copying of data when dereferencing the pointer, but due to the fact both point to the same underlying value array. if you look at the `big.Int` struct you'll see that the sign (`neg`) is separated from the actual value (`abs`), so both point to the same underlying value array, but the sign is not mutated when you do the `SetInt64` because this value is actually copied and not manipulated by the `Set`.